File: polymarket/trading/market_maker.py
"""
Market-making strategies for weather prediction markets
"""

import logging
import numpy as np
import pandas as pd
from typing import Dict, Tuple

from polymarket.config import MONTE_CARLO_PARAMS

logger = logging.getLogger(__name__)


class MarketMaker:
    """
    Market maker for Polymarket weather derivatives
    Manages bid-ask spreads and position sizing
    """

    def __init__(self, capital: float = 100000):
        self.capital = capital
        self.inventory = {}  # Track inventory by market
        self.pnl = 0.0

    def calculate_fair_value_spread(
        self,
        market_price: float,
        our_estimate: float,
        volatility: float,
        inventory: float = 0
    ) -> Dict[str, float]:
        """
        Calculate optimal bid-ask spread based on fair value and inventory

        Args:
            market_price: Current market mid price
            our_estimate: Our probability estimate (fair value)
            volatility: Market volatility
            inventory: Current inventory position

        Returns:
            Dictionary with bid and ask prices
        """
        # Base spread component (proportional to volatility)
        base_spread = volatility * 0.02  # 2% of volatility

        # Inventory component (penalize adverse positions)
        inventory_spread = max(0, abs(inventory) * 0.01)

        # Fair value component
        fair_value = our_estimate
        mid_price = market_price

        # Optimal spread
        half_spread = base_spread + inventory_spread

        # Asymmetric spread based on fair value
        if fair_value > mid_price:
            # We think price should go up, make bid tighter
            bid = mid_price - half_spread * 0.5
            ask = mid_price + half_spread * 1.5
        elif fair_value < mid_price:
            # We think price should go down, make ask tighter
            bid = mid_price - half_spread * 1.5
            ask = mid_price + half_spread * 0.5
        else:
            # Symmetric spread around market
            bid = mid_price - half_spread
            ask = mid_price + half_spread

        return {
            "bid": np.clip(bid, 0, 1),
            "ask": np.clip(ask, 0, 1),
            "spread": ask - bid,
            "mid": (bid + ask) / 2,
        }

    def optimal_bid_ask_spreads(
        self,
        market_prices: pd.DataFrame,
        probability_estimates: pd.Series,
        volatility: float = 0.15,
        inventory_limits: Dict[str, float] = None
    ) -> pd.DataFrame:
        """
        Calculate optimal spreads for a basket of markets

        Args:
            market_prices: DataFrame with market prices (columns: bid, ask, mid)
            probability_estimates: Series with our probability estimates
            volatility: Overall market volatility
            inventory_limits: Limits on inventory by market

        Returns:
            DataFrame with bid-ask recommendations
        """
        spreads = []

        for market_id in market_prices.index:
            market = market_prices.loc[market_id]
            estimate = probability_estimates.loc[market_id]
            inventory = self.inventory.get(market_id, 0)

            spread = self.calculate_fair_value_spread(
                market_price=market.get("mid", (market["bid"] + market["ask"]) / 2),
                our_estimate=estimate,
                volatility=volatility,
                inventory=inventory
            )

            spread["market_id"] = market_id
            spread["current_inventory"] = inventory
            spreads.append(spread)

        return pd.DataFrame(spreads).set_index("market_id")

    def calculate_position_sizes(
        self,
        probability_estimates: Dict[str, float],
        market_prices: Dict[str, float],
        volatility_estimates: Dict[str, float],
        max_position_pct: float = 0.1
    ) -> Dict[str, float]:
        """
        Calculate optimal position sizes using Kelly Criterion

        Args:
            probability_estimates: Probability estimates for each market
            market_prices: Current market prices
            volatility_estimates: Volatility estimates
            max_position_pct: Maximum position as % of capital

        Returns:
            Dictionary with recommended position sizes
        """
        position_sizes = {}

        for market_id, estimate in probability_estimates.items():
            market_price = market_prices.get(market_id, 0.5)
            volatility = volatility_estimates.get(market_id, 0.15)

            # Win probability
            p_win = estimate if estimate > market_price else 1 - estimate
            p_loss = 1 - p_win

            # Kelly Criterion: f* = (bp - q) / b
            # b = odds, p = probability, q = 1-p
            b = 1 / market_price - 1 if market_price > 0 else 1
            kelly_fraction = (b * p_win - p_loss) / b if b > 0 else 0

            # Apply fractional Kelly for safety (1/4 Kelly)
            kelly_fraction = max(0, kelly_fraction * 0.25)

            # Apply position size limit
            position_size = min(kelly_fraction, max_position_pct)

            position_sizes[market_id] = position_size

        return position_sizes

    def execute_market_orders(
        self,
        market_id: str,
        side: str,  # BUY or SELL
        size: float,
        price: float
    ) -> Dict:
        """
        Execute a market order

        Args:
            market_id: Market identifier
            side: BUY or SELL
            size: Position size
            price: Execution price

        Returns:
            Trade execution record
        """
        notional = size * price * 100  # 100x leverage

        if side == "BUY":
            if notional > self.capital:
                logger.warning(f"Insufficient capital for {notional} order")
                size = self.capital / (price * 100)
            self.inventory[market_id] = self.inventory.get(market_id, 0) + size
        elif side == "SELL":
            self.inventory[market_id] = self.inventory.get(market_id, 0) - size

        trade = {
            "market_id": market_id,
            "side": side,
            "size": size,
            "price": price,
            "notional": size * price * 100,
            "timestamp": pd.Timestamp.now(),
        }

        return trade

    def hedge_position(
        self,
        market_id: str,
        inventory: float,
        current_price: float,
        hedge_ratio: float = 0.5
    ) -> Dict:
        """
        Hedge an adverse inventory position

        Args:
            market_id: Market to hedge
            inventory: Current inventory
            current_price: Current market price
            hedge_ratio: Ratio to hedge (0-1)

        Returns:
            Hedge trade record
        """
        if inventory > 0:
            # Sell to reduce long position
            hedge_size = inventory * hedge_ratio
            return self.execute_market_orders(market_id, "SELL", hedge_size, current_price)
        else:
            # Buy to reduce short position
            hedge_size = abs(inventory) * hedge_ratio
            return self.execute_market_orders(market_id, "BUY", hedge_size, current_price)

    def calculate_pnl_by_market(
        self,
        current_prices: Dict[str, float]
    ) -> Dict[str, float]:
        """
        Calculate P&L for each market position

        Args:
            current_prices: Current market prices

        Returns:
            Dictionary with P&L by market
        """
        pnl_by_market = {}

        for market_id, inventory in self.inventory.items():
            if inventory != 0:
                current_price = current_prices.get(market_id, 0.5)
                # Simplified P&L calculation
                pnl = inventory * (current_price - 0.5) * 100

                pnl_by_market[market_id] = pnl

        return pnl_by_market

    def get_portfolio_metrics(self) -> Dict:
        """Calculate overall portfolio metrics"""
        total_position = sum(abs(v) for v in self.inventory.values())
        net_exposure = sum(self.inventory.values())

        return {
            "total_position": total_position,
            "net_exposure": net_exposure,
            "number_of_markets": len([m for m in self.inventory.values() if m != 0]),
            "total_pnl": self.pnl,
            "remaining_capital": self.capital - abs(self.pnl),
        }

File: src/trading/market_maker.rs
use std::collections::HashMap;

use chrono::Utc;

use crate::types::{MarketQuote, SpreadQuote};
use crate::utils::clip;

#[derive(Debug, Clone)]
pub struct TradeExecution {
    pub market_id: String,
    pub side: String,
    pub size: f64,
    pub price: f64,
    pub notional: f64,
    pub timestamp: chrono::DateTime<Utc>,
}

#[derive(Debug, Clone)]
pub struct PortfolioMetrics {
    pub total_position: f64,
    pub net_exposure: f64,
    pub number_of_markets: usize,
    pub total_pnl: f64,
    pub remaining_capital: f64,
}

#[derive(Debug, Clone)]
pub struct MarketMaker {
    pub capital: f64,
    pub inventory: HashMap<String, f64>,
    pub pnl: f64,
}

impl MarketMaker {
    pub fn new(capital: f64) -> Self {
        Self {
            capital,
            inventory: HashMap::new(),
            pnl: 0.0,
        }
    }

    pub fn calculate_fair_value_spread(
        &self,
        market_price: f64,
        our_estimate: f64,
        volatility: f64,
        inventory: f64,
    ) -> SpreadQuote {
        let base_spread = volatility * 0.02;
        let inventory_spread = inventory.abs() * 0.01;
        let half_spread = base_spread + inventory_spread;

        let (bid, ask) = if our_estimate > market_price {
            (market_price - half_spread * 0.5, market_price + half_spread * 1.5)
        } else if our_estimate < market_price {
            (market_price - half_spread * 1.5, market_price + half_spread * 0.5)
        } else {
            (market_price - half_spread, market_price + half_spread)
        };

        let bid = clip(bid, 0.0, 1.0);
        let ask = clip(ask, 0.0, 1.0);

        SpreadQuote {
            bid,
            ask,
            spread: ask - bid,
            mid: (bid + ask) / 2.0,
            market_id: String::new(),
            current_inventory: inventory,
        }
    }

    pub fn optimal_bid_ask_spreads(
        &self,
        market_prices: &HashMap<String, MarketQuote>,
        probability_estimates: &HashMap<String, f64>,
        volatility: f64,
    ) -> HashMap<String, SpreadQuote> {
        let mut out = HashMap::new();

        for (market_id, quote) in market_prices {
            let estimate = *probability_estimates.get(market_id).unwrap_or(&0.5);
            let inventory = *self.inventory.get(market_id).unwrap_or(&0.0);
            let mid = quote.mid.unwrap_or((quote.bid + quote.ask) / 2.0);

            let mut spread = self.calculate_fair_value_spread(mid, estimate, volatility, inventory);
            spread.market_id = market_id.clone();
            spread.current_inventory = inventory;
            out.insert(market_id.clone(), spread);
        }

        out
    }

    pub fn calculate_position_sizes(
        &self,
        probability_estimates: &HashMap<String, f64>,
        market_prices: &HashMap<String, f64>,
        max_position_pct: f64,
    ) -> HashMap<String, f64> {
        let mut sizes = HashMap::new();

        for (market_id, estimate) in probability_estimates {
            let market_price = *market_prices.get(market_id).unwrap_or(&0.5);
            let p_win = if *estimate > market_price {
                *estimate
            } else {
                1.0 - *estimate
            };
            let p_loss = 1.0 - p_win;

            let b = if market_price > 0.0 {
                1.0 / market_price - 1.0
            } else {
                1.0
            };
            let kelly = if b > 0.0 {
                ((b * p_win - p_loss) / b).max(0.0) * 0.25
            } else {
                0.0
            };

            sizes.insert(market_id.clone(), kelly.min(max_position_pct));
        }

        sizes
    }

    pub fn execute_market_orders(
        &mut self,
        market_id: &str,
        side: &str,
        size: f64,
        price: f64,
    ) -> TradeExecution {
        let mut adjusted_size = size;
        let notional = size * price * 100.0;

        if side == "BUY" {
            if notional > self.capital && price > 0.0 {
                adjusted_size = self.capital / (price * 100.0);
            }
            *self.inventory.entry(market_id.to_string()).or_insert(0.0) += adjusted_size;
        } else if side == "SELL" {
            *self.inventory.entry(market_id.to_string()).or_insert(0.0) -= adjusted_size;
        }

        TradeExecution {
            market_id: market_id.to_string(),
            side: side.to_string(),
            size: adjusted_size,
            price,
            notional: adjusted_size * price * 100.0,
            timestamp: Utc::now(),
        }
    }

    pub fn hedge_position(
        &mut self,
        market_id: &str,
        inventory: f64,
        current_price: f64,
        hedge_ratio: f64,
    ) -> TradeExecution {
        if inventory > 0.0 {
            self.execute_market_orders(market_id, "SELL", inventory * hedge_ratio, current_price)
        } else {
            self.execute_market_orders(
                market_id,
                "BUY",
                inventory.abs() * hedge_ratio,
                current_price,
            )
        }
    }

    pub fn calculate_pnl_by_market(&self, current_prices: &HashMap<String, f64>) -> HashMap<String, f64> {
        let mut pnl = HashMap::new();

        for (market_id, inventory) in &self.inventory {
            if *inventory != 0.0 {
                let current = *current_prices.get(market_id).unwrap_or(&0.5);
                pnl.insert(market_id.clone(), inventory * (current - 0.5) * 100.0);
            }
        }

        pnl
    }

    pub fn get_portfolio_metrics(&self) -> PortfolioMetrics {
        let total_position = self.inventory.values().map(|v| v.abs()).sum::<f64>();
        let net_exposure = self.inventory.values().sum::<f64>();
        let number_of_markets = self.inventory.values().filter(|v| **v != 0.0).count();

        PortfolioMetrics {
            total_position,
            net_exposure,
            number_of_markets,
            total_pnl: self.pnl,
            remaining_capital: self.capital - self.pnl.abs(),
        }
    }
}
